server/py/framework/db/sqldb/db.py

```diff
@@ -7149,3 +7152,103 @@ def _paginate_query(
             query = query.limit(limit)
 
         return query
+
+    @staticmethod
+    def _find_artifact_latest_tag(session, object_record):
+        """Find the 'latest' tag for an artifact object."""
+        return (
+            session.query(ArtifactV2.Tag)
+            .filter(
+                ArtifactV2.Tag.obj_id == object_record.id,
+                ArtifactV2.Tag.name == mlrun.common.constants.RESERVED_TAG_NAME_LATEST,
+                ArtifactV2.Tag.project == object_record.project,
+            )
+            .one_or_none()
+        )
+
+    @staticmethod
+    def _find_other_latest_in_other_iterations(session, object_record):
+        """Check if there are other 'latest' tags in other iterations for the same producer_id."""
+        return (
+            session.query(ArtifactV2.Tag)
+            .filter(
+                ArtifactV2.Tag.obj_id != object_record.id,
+                ArtifactV2.Tag.name == mlrun.common.constants.RESERVED_TAG_NAME_LATEST,
+                ArtifactV2.Tag.project == object_record.project,
+                ArtifactV2.Tag.obj_name == object_record.key,
+                ArtifactV2.Tag.obj_id.in_(
+                    session.query(ArtifactV2.id).filter(
+                        ArtifactV2.producer_id == object_record.producer_id,
+                        ArtifactV2.iteration != object_record.iteration,
+                    )
+                ),
+            )
+            .first()
+        )
+
+    @staticmethod
+    def _find_previous_most_recent_artifact(session, object_record):
+        """Find the most recent artifact based on the update timestamp, excluding the current artifact."""
+        query = session.query(ArtifactV2).filter(
```

Comment:
You don't really care for all the artifact object here, but just the `id`. So you should use the `with_entities(ArtifactV2.id)` option to get back just the `id` of the object, rather than the full object (which contains the entire JSON structure etc.). It will generate a small performance gain.

Reply:
Fixed